src/context.cpp

```diff
@@ -48,4 +54,72 @@ namespace powerloader
             spdlog::set_level(spdlog::level::warn);
         }
     }
+
+
+    mirror_set mirror_map_type::get_mirrors(const std::string& host_name) const
+    {
+        auto find_it = find(host_name);
+        if (find_it == end())
+            return {};
+
+        return find_it->second;
+    }
+
+    // Returns true if there are registered mirrors stored here, false if none are.
+    bool mirror_map_type::has_mirrors(const std::string& host_name) const
+    {
+        auto find_it = find(host_name);
+        return find_it != end() && !find_it->second.empty();
+    }
+
+    bool mirror_map_type::add_unique_mirror(const std::string& host_name,
+                                            std::shared_ptr<Mirror> mirror)
+    {
+        auto find_it = find(host_name);
+        if (find_it != end())
+        {
+            auto& mirrors = find_it->second;
+            if (details::already_exists(mirror->id(), mirrors))
+                return false;
+            mirrors.push_back(std::move(mirror));
+        }
+        else
+        {
+            (*this)[host_name] = { std::move(mirror) };
+        }
+        return true;
+    }
+
+    void mirror_map_type::reset(mirror_map_base new_values)
+    {
+        if (details::is_every_mirror_unique_per_host(new_values))
+            throw std::invalid_argument("mirror map must have unique mirrors per host name");
+        static_cast<mirror_map_base&>(*this) = std::move(new_values);
+    }
+
+    namespace details
+    {
+        bool already_exists(const MirrorID& id, const mirror_set& mirrors)
+        {
+            for (auto&& mirror : mirrors)
+                if (mirror->id() == id)
+                    return true;
+            return false;
+        }
+
+        bool is_every_mirror_unique_per_host(const mirror_map_base& mirrors)
+        {
+            std::set<MirrorID> mirrors_ids;  // TODO: replace by flat_set once available.
+            for (const auto& slot : mirrors)
+            {
+                for (const auto& mirror : slot.second)
+                {
+                    auto [_, success] = mirrors_ids.insert(mirror->id());
+                    if (!success)
+                        return false;
+                }
+            }
+            return true;
```

Comment:
I'm confused with this method, the name suggests it checks if the mirrors are unique per host, but the implementation checks if the mirror ids are unique across the whole map. Shouldn't the implementation start with:

```cpp
for (const auto& slot : mirrors)
{
    std::set<MirrorID> mirrors_ids;  // TODO: replace by flat_set once available.
    // ...
}
```

Reply:
woooops, you're right, good thing I wrote the line specifying what it should do XD